Zend/zend_compile.c

```diff
@@ -5158,6 +5158,244 @@ void zend_compile_switch(zend_ast *ast) /* {{{ */
+			Z_TRY_ADDREF_P(CT_CONSTANT(opline->op1));
+		}
+		opnum_match = opline - CG(active_op_array)->opcodes;
+	}
+
+	uint32_t cond_count = 0;
+	uint32_t *jmpnz_opnums = safe_emalloc(sizeof(uint32_t), num_conds, 0);
+	for (uint32_t i = 0; i < arms->children; ++i) {
+		zend_ast *arm_ast = arms->child[i];
+
+		if (!arm_ast->child[0]) {
+			if (has_default_arm) {
+				CG(zend_lineno) = arm_ast->lineno;
+				zend_error_noreturn(E_COMPILE_ERROR,
+					"Match expressions may only contain one default arm");
+			}
+			has_default_arm = 1;
+			continue;
+		}
+
+		zend_ast_list *conds = zend_ast_get_list(arm_ast->child[0]);
+		for (uint32_t j = 0; j < conds->children; j++) {
+			zend_ast *cond_ast = conds->child[j];
+
+			znode cond_node;
+			zend_compile_expr(&cond_node, cond_ast);
+
+			if (expr_node.op_type == IS_CONST
+				&& Z_TYPE(expr_node.u.constant) == IS_FALSE) {
+				jmpnz_opnums[cond_count] = zend_emit_cond_jump(ZEND_JMPZ, &cond_node, 0);
+			} else if (expr_node.op_type == IS_CONST
+				&& Z_TYPE(expr_node.u.constant) == IS_TRUE) {
+				jmpnz_opnums[cond_count] = zend_emit_cond_jump(ZEND_JMPNZ, &cond_node, 0);
+			} else {
+				zend_op *opline = zend_emit_op(NULL, ZEND_IS_IDENTICAL, &expr_node, &cond_node);
+				SET_NODE(opline->result, &case_node);
+				if (opline->op1_type == IS_CONST) {
+					Z_TRY_ADDREF_P(CT_CONSTANT(opline->op1));
+				}
+
+				jmpnz_opnums[cond_count] = zend_emit_cond_jump(ZEND_JMPNZ, &case_node, 0);
+			}
+
+			cond_count++;
+		}
+	}
+
+	uint32_t opnum_default_jmp = zend_emit_jump(0);
+	zend_bool is_first_case = 1;
+	cond_count = 0;
+
+	for (uint32_t i = 0; i < arms->children; ++i) {
+		zend_ast *arm_ast = arms->child[i];
+		zend_ast *stmt_ast = arm_ast->child[1];
+
+		if (arm_ast->child[0] != NULL) {
+			zend_ast_list *conds = zend_ast_get_list(arm_ast->child[0]);
+
+			for (uint32_t j = 0; j < conds->children; j++) {
+				zend_ast *cond_ast = conds->child[j];
+				zend_update_jump_target_to_next(jmpnz_opnums[cond_count]);
+
+				if (jumptable) {
+					zval *cond_zv = zend_ast_get_zval(cond_ast);
+					zval jmp_target;
+					ZVAL_LONG(&jmp_target, get_next_op_number());
+
+					ZEND_ASSERT(Z_TYPE_P(cond_zv) == jumptable_type);
+					if (Z_TYPE_P(cond_zv) == IS_LONG) {
+						zend_hash_index_add(jumptable, Z_LVAL_P(cond_zv), &jmp_target);
+					} else {
+						ZEND_ASSERT(Z_TYPE_P(cond_zv) == IS_STRING);
+						zend_hash_add(jumptable, Z_STR_P(cond_zv), &jmp_target);
+					}
+				}
+
+				cond_count++;
+			}
+		} else {
+			zend_update_jump_target_to_next(opnum_default_jmp);
+
+			if (jumptable) {
+				ZEND_ASSERT(opnum_match != (uint32_t)-1);
+				zend_op *opline = &CG(active_op_array)->opcodes[opnum_match];
+				opline->extended_value = get_next_op_number();
+			}
+		}
+
+		if (result_unused) {
+			zend_compile_stmt(stmt_ast);
+		} else {
+			if (stmt_ast->kind == ZEND_AST_STMT_LIST) {
+				zend_error_noreturn(E_COMPILE_ERROR, "Match expressions that utilize the result value can't contain statement lists");
```

Comment:
nit: `use` is much, much more common in PHP's error messages (half of those lines are in places "utilize" could be used)

e.g. `zend_error_noreturn(E_COMPILE_ERROR, "Cannot use result of built-in function in write context");`

```
» ag 'zend_error.*\buse\b'|wc
38
» ag 'zend_error.*\butilize\b'|wc -l
0
```

Reply:
It's also technically incorrect. It's not the match not the `match` expression that uses the expression result. What do you think about:

> Match expressions whose return value is used value can't contain statement lists

Reply:
Oh, I wasn't looking at the grammar of the message.

Match expressions whose return value is used ~~value~~ can't contain statement lists

Maybe "where the" instead of "whose" would make more sense, since expressions are things

Reply:
Also, the error is a bit misleading - for example, `for (match(...) {...}; ; )` is technically not using the return value of match, but can't contain statement lists in the current implementation.

Maybe `Match expressions that aren't statements can't contain statement lists`

Reply:
> Maybe "where the" instead of "whose" would make more sense, since expressions are things

I wasn't sure about that since I'm not a native English speaker. 🙂 That works for me.

Reply:
> technically not using the return value of match, but can't contain statement lists in the current implementation.

I've been struggling to find the right words to express this. Do you have any suggestions? There's also a lot of confusion around this in the mailing list.

Reply:
I haven't come up with a good error message either

Maybe `Match expressions that aren't statements can't contain statement lists`
Maybe `Statement lists are forbidden when match isn't used as a statement`

Reply:
I changed it to

> Match expressions that are not statements can't contain statement lists

for now. If anyone has another suggestion let me now 🙂